### PR TITLE
Intrinsic: introduce minimumnum and maximumnum

### LINKED ISSUE #93033: llvm.minnum  should be lowered to fminimum_numf instead of fminf
In https://llvm.org/docs/LangRef.html#llvm-minnum-intrinsic
```
Unlike the IEEE-754 2008 behavior, this does not distinguish between signaling and quiet NaN inputs. If a target’s implementation follows the standard and returns a quiet NaN if either input is a signaling NaN, the intrinsic lowering is responsible for quieting the inputs to correctly return the non-NaN input (e.g. by using the equivalent of llvm.canonicalize).
```

While in https://www.gnu.org/software/libc/manual/html_node/Misc-FP-Arithmetic.html
```
If an argument is a quiet NaN, the other argument is returned. If both arguments are NaN, or either is a signaling NaN, NaN is returned.
```

### PR DESCRIPTION
**Pull request by wzssyqa**
Currently, on different platform, the behaivor of llvm.minnum is different if one operand is sNaN:

When we compare sNaN vs NUM:

ARM/AArch64/PowerPC: follow the IEEE754-2008's minNUM: return qNaN. RISC-V/Hexagon follow the IEEE754-2019's minimumNumber: return NUM. X86: Returns NUM but not same with IEEE754-2019's minimumNumber as
     +0.0 is not always greater than -0.0.
MIPS/LoongArch/Generic: return NUM.
LIBCALL: returns qNaN.

So, let's introduce llvm.minmumnum/llvm.maximumnum, which always follow IEEE754-2019's minimumNumber/maximumNumber.

Half-fix: #93033

### CODE REVIEW

**Comment by wzssyqa** [comment]
Since we need to reuse some logic of `minnum/maxnum` to implement `minimumnum/maximumnum`,
let's add them before switch the behavior of `minnum/maxnum`.


Known not working ports, will be fixed in future PRs:

1. X86: the current `minnum/maxnum` cannot process +0 vs -0 as `minimumnum/maximumnum` expected.
2. ARM(32): has some interaction with the behavior of `minnum/maxnum`.
3. PowerPC: has some interaction with the behavior of `minnum/maxnum`: need define `fcanonicalize`.

**Comment by efriedma-quic** [comment]
Please add a table to LangRef comparing the behavior of the three versions of min/max intrinsics for various inputs.

**Comment by wzssyqa** [comment]
@peterwaller-arm I noticed that in `llvm/test/CodeGen/AArch64/combine_andor_with_cmps.ll`, `FMAXNUM_IEEE` is claimed that it is not supported. While I noticed that `fmaxnm` follows the rules of `maxNUM` of IEEE754-2008.
Is there any other limitation of `fmaxnm`?

**Comment by wzssyqa** [comment]
TODO: implement for architectures that don't have `fmin/fmax` instructions:

This is the example of MIPS pre-R6:

```
mins:
        .set    noreorder
        .set    nomacro
        mtc1    $0,$f1
        add.s   $f0,$f12,$f1
        add.s   $f13,$f13,$f1

        c.un.s  $fcc0,$f0,$f0
        movt.s  $f0,$f13,$fcc0
        c.un.s  $fcc0,$f13,$f13
        movt.s  $f13,$f0,$fcc0
        c.ult.s $fcc0,$f13,$f0
        movt.s  $f0,$f13,$fcc0
        jr      $31
```

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -5005,8 +5007,11 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
       // If either operand is not NaN, the result is not NaN.
       if (NeverNaN && (IID == Intrinsic::minnum || IID == Intrinsic::maxnum))
         Known.knownNot(fcNan);
+      if (NeverNaN &&
+          (IID == Intrinsic::minimumnum || IID == Intrinsic::maximumnum))
+        Known.knownNot(fcNan);
 
-      if (IID == Intrinsic::maxnum) {
+      if (IID == Intrinsic::maxnum || IID == Intrinsic::maximumnum) {
```
Comment (arsenm) [inline llvm/lib/Analysis/ValueTracking.cpp:5014]: Best to keep the value tracking handling in a separate PR. This is missing test coverage

**Review by arsenm — COMMENTED** [review]
> 3. PowerPC: has some interaction with the behavior of `minnum/maxnum`: need define `fcanonicalize`.

AMDGPU has the same handling. This is to break the signaling nan handling from IEEE to the broken old glibc libm behavior. If we fix the definition to match IEEE, this is no longer necessary and the operation is directly legal

**Comment by peterwaller-arm** [comment]
> @peterwaller-arm I noticed that in `llvm/test/CodeGen/AArch64/combine_andor_with_cmps.ll`, `FMAXNUM_IEEE` is claimed that it is not supported. While I noticed that `fmaxnm` follows the rules of `maxNUM` of IEEE754-2008. Is there any other limitation of `fmaxnm`?

This is not an area I know a great deal about and I'm about to go on leave, so I'm probably not the best person to ask. Perhaps the author of the comment (on https://reviews.llvm.org/D159240) @kmitropoulou can help? Could the comment mean that it's not currently supported by LLVM rather than not supported by the ISA? I see that the details of how NaNs are handled matters. I'm afraid I don't have time to get to the bottom of this currently.

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -16049,6 +16094,84 @@ of the two arguments. -0.0 is considered to be less than +0.0 for this
...
+""""""""""
+
+The arguments and return value are floating-point numbers of the same
+type.
+
+Semantics:
+""""""""""
+If both operands are NaNs, returns qNaN. Otherwise returns the lesser
+of the two arguments. -0.0 is considered to be less than +0.0 for this
+intrinsic. Note that these are the semantics specified in IEEE 754-2019.
```
[the remark this answers is not in the document]
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16134]: > Needs to spell out the signaling nan behavior. If we're fixing minnum's snan behavior to match IEEE, this is identical except with the stronger guarantee for signed zero ordering. The documentation should also explicitly state this is the only difference, to help reduce confusion. Alternatively, we could add an immediate bool parameter to minnum/maxnum for whether the ordering of 0 is guaranteed
> 

OK. I will add more details here.

> I hate the naming mess we've ended up with here, but I guess C23 has damned us. If you're going to match the C23 names, this should be `llvm.minimum.num` with an extra _

In fact, in my initial version of patch, I use _, then I get `llvm.minimum.num`. I considered it as a problem, since it cost some time to know in fact I should use `llvm.minimum.num` instead of `llvm.minimum_num` :(

**Comment by wzssyqa** [comment]
> > 3. PowerPC: has some interaction with the behavior of `minnum/maxnum`: need define `fcanonicalize`.
> 
> AMDGPU has the same handling. This is to break the signaling nan handling from IEEE to the broken old glibc libm behavior. If we fix the definition to match IEEE, this is no longer necessary and the operation is directly legal

Added to TODO list.

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -15868,6 +15868,51 @@ The returned value is completely identical to the input except for the sign bit;
...
+
+   * - Operation
+     - minnum/maxnum
+     - minimum/maximum
+     - minimumnum/maximumnum
+
+   * - ``NUM vs qNaN``
+     - NUM, no exception
+     - qNaN, no exception
+     - qNaN, no exception
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:15888]: This line is not correct. (`minimumnum(x, qNaN)` is `x`)

File: llvm/docs/LangRef.rst [hunk 4]
```diff
@@ -15868,6 +15868,51 @@ The returned value is completely identical to the input except for the sign bit;
...
+
+   * - ``+0.0 vs -0.0``
+     - either one
+     - +0.0(max)/-0.0(min)
+     - +0.0(max)/-0.0(min)
+
+   * - ``NUM vs NUM``
+     - larger(max)/smaller(min)
+     - larger(max)/smaller(min)
+     - larger(max)/smaller(min)
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:15913]: It feels like this section needs a callout on how the sNaN entries interact with our current NaN-handling policy.

File: llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h [hunk 5]
```diff
@@ -1723,6 +1723,18 @@ class MachineIRBuilder {
     return buildInstr(TargetOpcode::G_FMAXNUM_IEEE, {Dst}, {Src0, Src1}, Flags);
   }
 
+  MachineInstrBuilder
+  buildFMinimumNUM(const DstOp &Dst, const SrcOp &Src0, const SrcOp &Src1,
```
Comment (jcranmer-intel) [inline llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h:1727]: How about `FMinimumnum` instead?

File: llvm/include/llvm/IR/Intrinsics.td [hunk 6]
```diff
@@ -1275,6 +1283,14 @@ let IntrProperties = [IntrInaccessibleMemOnly, IntrWillReturn, IntrStrictFP] in
...
                                                          LLVMMatchType<0>,
                                                          llvm_metadata_ty ]>;
+  def int_experimental_constrained_maximumnum : DefaultAttrsIntrinsic<[ llvm_anyfloat_ty ],
+                                                       [ LLVMMatchType<0>,
+                                                         LLVMMatchType<0>,
+                                                         llvm_metadata_ty ]>;
+  def int_experimental_constrained_minimumnum : DefaultAttrsIntrinsic<[ llvm_anyfloat_ty ],
+                                                       [ LLVMMatchType<0>,
+                                                         LLVMMatchType<0>,
+                                                         llvm_metadata_ty ]>;
```
Comment (jcranmer-intel) [inline llvm/include/llvm/IR/Intrinsics.td:1293]: You still need documentation for all these extra intrinsics.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 7]
```diff
@@ -9130,6 +9142,15 @@ void SelectionDAGBuilder::visitCall(const CallInst &I) {
         if (visitBinaryFloatCall(I, ISD::FMAXNUM))
           return;
         break;
+      case LibFunc_fminimum_num:
+      case LibFunc_fminimum_numf:
+        if (visitBinaryFloatCall(I, ISD::FMINIMUMNUM))
+          return;
+        break;
+      case LibFunc_fmaximum_num:
```
Comment (jcranmer-intel) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:9150]: Methinks there's quite a few missing case statements here.

File: llvm/unittests/ADT/APFloatTest.cpp [hunk 8]
```diff
@@ -631,6 +631,46 @@ TEST(APFloatTest, Maximum) {
...
+  APFloat f2(2.0);
+  APFloat zp(0.0);
+  APFloat zn(-0.0);
+  APFloat nan = APFloat::getNaN(APFloat::IEEEdouble());
+  APFloat snan = APFloat::getSNaN(APFloat::IEEEdouble());
+
+  EXPECT_EQ(1.0, minimumnum(f1, f2).convertToDouble());
+  EXPECT_EQ(1.0, minimumnum(f2, f1).convertToDouble());
+  EXPECT_EQ(-0.0, minimumnum(zp, zn).convertToDouble());
+  EXPECT_EQ(-0.0, minimumnum(zn, zp).convertToDouble());
```
Comment (jcranmer-intel) [inline llvm/unittests/ADT/APFloatTest.cpp:645]: I don't think this is testing what you want to test, as `0.0 == -0.0`

Comment (wzssyqa) [reply llvm/unittests/ADT/APFloatTest.cpp:645]: Let's add some new isNegtive tests.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp (hunk 7 above)
Comment (wzssyqa) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:9150]: LibFunc_fminimum_numl
LibFunc_fmaximum_numf
LibFunc_fmaximum_numl

Add. Should we add something like `fmaximum_numfN` and `fmaximum_numfNx `?

Comment (wzssyqa) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:9151]: If need, I think that we need to a new patch, since neither other  libcalls have them.

File: llvm/include/llvm/Analysis/IVDescriptors.h [hunk 9]
```diff
@@ -32,27 +32,29 @@ class StoreInst;
...
+  UMin,        ///< Unsigned integer min implemented in terms of select(cmp()).
+  UMax,        ///< Unsigned integer max implemented in terms of select(cmp()).
+  FAdd,        ///< Sum of floats.
+  FMul,        ///< Product of floats.
+  FMin,        ///< FP min implemented in terms of select(cmp()).
+  FMax,        ///< FP max implemented in terms of select(cmp()).
+  FMinimum,    ///< FP min with llvm.minimum semantics
+  FMaximum,    ///< FP max with llvm.maximum semantics
+  FMinimumnum, ///< FP min with llvm.minimumnum semantics
+  FMaximumnum, ///< FP max with llvm.maximumnum semantics
```
Comment (arsenm) [inline llvm/include/llvm/Analysis/IVDescriptors.h:52]: Not sure this is tested, but updating each optimization should be split into a separate change

Comment (wzssyqa) [reply llvm/include/llvm/Analysis/IVDescriptors.h]: Removed from this PR, and add them into TODOs.

File: llvm/docs/LangRef.rst [hunk 10]
```diff
@@ -16055,6 +16145,90 @@ of the two arguments. -0.0 is considered to be less than +0.0 for this
...
+type.
+
+Semantics:
+""""""""""
+If both operands are NaNs (including sNaN), returns qNaN. If one operand
+is NaN (including sNaN) and another operand is a number, return the number.
+Otherwise returns the lesser of the two arguments. -0.0 is considered to
+be less than +0.0 for this intrinsic.
+
+Note that these are the semantics of minimumNumber specified in IEEE 754-2019.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16188]: This is not the IEEE semantics for signaling nan. For a signaling nan, returns a quiet nan, not the non-non operand

File: llvm/docs/LangRef.rst [hunk 11]
```diff
@@ -16055,6 +16145,90 @@ of the two arguments. -0.0 is considered to be less than +0.0 for this
...
+type.
+
+Semantics:
+""""""""""
+If both operands are NaNs (including sNaN), returns qNaN. If one operand
+is NaN (including sNaN) and another operand is a number, return the number.
+Otherwise returns the greater of the two arguments. -0.0 is considered to
+be less than +0.0 for this intrinsic.
+
+Note that these are the semantics of minimumNumber specified in IEEE 754-2019.
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:16230]: Copy paste error minimumNumber. Also like above, this is not the signaling nan behavior (where the behavior inverts from quiet nan)

Comment (arsenm) [reply llvm/docs/LangRef.rst:16230]: Should also state the explicit difference here, on the intrinsic, the comparison to minnum

File: llvm/docs/LangRef.rst [hunk 12]
```diff
@@ -15874,6 +15874,96 @@ The returned value is completely identical to the input except for the sign bit;
...
+     - either one
+     - +0.0 > -0.0
+     - +0.0 > -0.0
+     - +0.0 > -0.0
+
+   * - ``NUM vs sNaN``
+     - qNaN, invalid excpetion
+     - qNaN, invalid excpetion
+     - qNaN, invalid excpetion
+     - NUM, invalid excpetion
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:15913]: Typo exception throughout

File: llvm/docs/LangRef.rst [hunk 13]
```diff
@@ -15874,6 +15874,96 @@ The returned value is completely identical to the input except for the sign bit;
...
+
+   * - ``NUM vs qNaN``
+     - NUM, no excpetion
+     - NUM, no excpetion
+     - qNaN, no excpetion
+     - NUM, no excpetion
+
```
Comment (arsenm) [inline llvm/docs/LangRef.rst:15920]: cover nan vs. nan case

File: llvm/docs/LangRef.rst (hunk 10 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:16188]: No. minimumNumber returns NaN only when both are NaN.

```
minimumNumber(x, y) is x if x < y, y if y < x, and the number if one operand is a number and the
other is a NaN. For this operation, −0 compares less than +0. If x = y and signs are the same it is
either x or y. If both operands are NaNs, a quiet NaN is returned, according to 6.2. If either
operand is a signaling NaN, an invalid operation exception is signaled, but unless both operands
are NaNs, the signaling NaN is otherwise ignored and not converted to a quiet NaN as stated in
6.2 for other operations.
```

File: llvm/include/llvm/ADT/APFloat.h [hunk 14]
```diff
@@ -1449,6 +1449,16 @@ inline APFloat minimum(const APFloat &A, const APFloat &B) {
     return A.isNegative() ? A : B;
   return B < A ? B : A;
 }
+LLVM_READONLY
```
Comment (kuhar) [inline llvm/include/llvm/ADT/APFloat.h:1452]: Please add and an empty line before this function and document its semantics.

File: llvm/include/llvm/ADT/APFloat.h [hunk 15]
```diff
@@ -1462,6 +1472,16 @@ inline APFloat maximum(const APFloat &A, const APFloat &B) {
     return A.isNegative() ? B : A;
   return A < B ? B : A;
 }
+LLVM_READONLY
```
Comment (kuhar) [inline llvm/include/llvm/ADT/APFloat.h:1475]: also here

File: llvm/docs/LangRef.rst [hunk 16]
```diff
@@ -15883,6 +15883,95 @@ The returned value is completely identical to the input except for the sign bit;
...
+   * - ``IEEE754``
+     - none
+     - nimNUM/maxNUM (2008)
+     - minimum/maximum (2019)
+     - minimumNumber/maximumNumber (2019)
+
+   * - ``+0.0 vs -0.0``
+     - either one
+     - +0.0 > -0.0
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:15914]: My copy of IEEE 754-2008 says of minNum:

> minNum(x, y) is the canonicalized number x if x<y, y if y<x, the canonicalized number if one
operand is a number and the other a quiet NaN. Otherwise it is either x or y, canonicalized (this
means results might differ among implementations). 

IEEE 754 doesn't guarantee the sign of `minNum(-0.0, +0.0)`, just like `fmin`.

Note also that TS 18611-1 bound C `fmin` to IEEE 754-2008 `minNum` (https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1711.pdf)--`fmin` and `minNum` should also be considered equivalent.

(Also, while I'm being pedantic, the spelling used in IEEE 754 is `minNum`, not `minNUM`.

File: llvm/docs/LangRef.rst [hunk 17]
```diff
@@ -15883,6 +15883,95 @@ The returned value is completely identical to the input except for the sign bit;
...
+
+   * - ``NUM/qNaN vs sNaN``
+     - qNaN, invalid exception
+     - qNaN, invalid exception
+     - qNaN, invalid exception
+     - NUM/qNaN, invalid exception
+
+   * - ``NUM/qNaN vs qNaN``
+     - NUM/qNaN, no excpetion
```
Comment (jcranmer-intel) [inline llvm/docs/LangRef.rst:15925]: EMISSPELLED in this column

File: llvm/docs/LangRef.rst (hunk 4 above)
Comment (jcranmer-intel) [reply llvm/docs/LangRef.rst:15964]: The current text still isn't covering the interaction of sNaN with LLVM's NaN-handling policy.

It's also not entirely accurate to say that invalid exceptions are raised when FP exceptions are unspecified without constrained intrinsics.

File: llvm/docs/LangRef.rst (hunk 16 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:15914]: Thanks. You are right.  I was confused by the hardware implementations of some hardware.
ARM/MIPSr6/PowerPC implement +0.0>-0.0.

File: llvm/docs/LangRef.rst (hunk 17 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:15925]: And I split it to `NUM vs qNaN` and `qNaN vs qNaN`.

Comment (wzssyqa) [reply llvm/docs/LangRef.rst:15925]: For more details about sNaN on constrained intrinsics, we may need to update them in future PRs.
Since this PR is only about add `minimumnum and maximumnum`.

File: llvm/docs/LangRef.rst (hunk 4 above)
Comment (wzssyqa) [reply llvm/docs/LangRef.rst:15964]: Sorry, I guess that I am not full understanding "LLVM's NaN-handling policy".
As my understanding, for fmax*, we should fellow the libc's semantics.

Comment (wzssyqa) [reply llvm/docs/LangRef.rst:15964]: For constrained ones, I guess that we should add more details in another PR(s).

File: llvm/docs/LangRef.rst [hunk 18]
```diff
@@ -15883,6 +15883,100 @@ The returned value is completely identical to the input except for the sign bit;
...
+.. list-table::
+   :header-rows: 2
+
+   * - ``ISO C``
+     - fmin/fmax
+     - fmininum/fmaximum
+     - fminimum_num/fmaximum_num
+
+   * - ``IEEE754``
+     - nimNum/maxNum (2008)
```
Comment (francisvm) [inline llvm/docs/LangRef.rst:15906]:
```suggestion
     - minNum/maxNum (2008)
```

File: llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir [hunk 19]
```diff
@@ -798,6 +804,12 @@
 # DEBUG-NEXT: .. opcode {{[0-9]+}} is aliased to {{[0-9]+}}
 # DEBUG-NEXT: .. type index coverage check SKIPPED: user-defined predicate detected
 # DEBUG-NEXT: .. imm index coverage check SKIPPED: user-defined predicate detected
+# DEBUG-NEXT: G_VECREDUCE_FMAXIMUMNUM (opcode 276): 2 type indices, 0 imm indices
+# DEBUG-NEXT: .. type index coverage check SKIPPED: no rules defined
+# DEBUG-NEXT: .. imm index coverage check SKIPPED: no rules defined
+# DEBUG-NEXT: G_VECREDUCE_FMINIMUMNUM (opcode 277): 2 type indices, 0 imm indices
```
Comment (dyung) [inline llvm/test/CodeGen/AArch64/GlobalISel/legalizer-info-validation.mir:810]: The opcodes for G_VECREDUCE_FMAXIMUMNUM and G_VECREDUCE_FMINIMUMNUM should be a regex to be consistent with the rest of the test and to make it more flexible.

**Comment by nikic** [comment]
It looks like this PR was merged without being approved, and I also couldn't find the corresponding RFC for this addition on discourse. I've reverted it for now.

**Comment by wzssyqa** [comment]
@nikic thanks. Please also revert https://github.com/llvm/llvm-project/commit/225d8fc8eb24fb797154c1ef6dcbe5ba033142da

**Comment by wzssyqa** [comment]
@nikic Thanks. I submit an RFC now
https://discourse.llvm.org/t/rfc-fix-llvm-min-f-and-llvm-max-f-intrinsics/79735
